README.md
# NAME

Imager::LineTrace - Line tracer

# SYNOPSIS

    # from Sample/bmp2figure.pl
    use Imager::LineTrace;

    my $img = Imager::LineTrace->new( file => $ARGV[0] ) or die Imager->errstr;
    my $figures_ref = $img->line_trace();

    my $i = 0;
    foreach my $figure (@{$figures_ref}) {
        print "-------- [", $i++, "] --------", "\n";
        print "type        : ", $figure->{type}, "\n";
        print "trace_value : ", sprintf("0x%06X", $figure->{value}), "\n";
        print "is_close: ", $figure->{is_closed}, "\n";
        foreach my $p (@{$figure->{points}}) {
            printf( "(%2d,%2d)\n", $p->[0], $p->[1] );
        }
    }

# DESCRIPTION

    # Tracing clockwise from left top.

Basic Overview

    my $img = Imager::LineTrace->new( file => $path ) or die Imager->errstr;

    # Trace black line on white.
    my $figures_ref = $img->line_trace();

    # If you want to trace counter-clockwise from left bottom.
    $img->filp( dir => 'v' );
    my $figures_ref = $img->line_trace();

    # If you want to select color. ( 0:R, 1:G, 2:B, 3:Alpha )
    my $figures_ref = $img->line_trace( channels => [0] );

    # Or you want to trace with R,G and B.(this is defalt.)
    my $figures_ref = $img->line_trace( channels => [0,1,2] );

    # If you want to trace not black color.
    my $figures_ref = $img->line_trace( ignore => 0 );

    # If you want to trace many figure. (default "limit" is 1024)
    my $figures_ref = $img->line_trace( limit => 10000 );

# LICENSE

Copyright (C) neko.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

neko <techno.cat.miau@gmail.com>


Changes
Revision history for Perl extension Imager-LineTrace

v0.05

  - Update POD.



lib/Imager/LineTrace/Figure.pm
package Imager::LineTrace::Figure;
use 5.008001;
use strict;
use warnings;

sub new {
    my $pkg = shift;
    my $args_ref = shift;

    my @points = map {
        bless $_, 'Imager::LineTrace::Point';
    } @{$args_ref->{points}};

    my $type = "Undefined";
    if ( $args_ref->{is_close} ) {
        $type = "Polygon";
    }
    elsif ( 3 <= scalar(@points) ) {
        $type = "Polyline";
    }
    elsif ( 2 <= scalar(@points) ) {
        $type = "Line";
    }
    else {
        $type = "Point";
    }

    bless {
        points    => \@points,
        is_closed => $args_ref->{is_closed},
        value     => $args_ref->{value},
        type      => $type
    }, $pkg;
}

1;
__END__

=encoding utf-8

=head1 NAME

Imager::LineTrace::Figure - Result of line trace

=head1 SYNOPSIS

    use Imager::LineTrace::Figure;

    my $figure = reverse (
        points    => [ [0, 0], [1, 0] ],
        is_closed => 0,
        value     => 1, # Traced value
    );

    print $figure->{type}; # => "Line"

=head1 DESCRIPTION

Result object of Imager::LineTracer.

RETURN DATA

Basic Overview

    # $figure->{points} is ARRAY reference.
    foreach my $point (@{$figure_ref->{points}}) {
        printf( "x = %d, y = %d\n", $point->[0], $point->[1] );
    }

    # If $figure->{is_closed} is 1, end point linked to start point.
    print $figure->{is_closed};

    # $figure->{value} is traced value.
    print $figure->{value};

    # $figure->{type} is one of "Point", "Line", "Polyline" and "Polygon".
    print $figure->{type};

=head1 LICENSE

Copyright (C) neko.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

neko E<lt>techno.cat.miau@gmail.comE<gt>

=cut


lib/Imager/LineTrace/Algorithm.pm
package Imager::LineTrace::Algorithm;
use 5.008001;
use strict;
use warnings;

sub search {
    my ( $pixels_ref, $opt ) = @_;
    my $ignore = ( exists $opt->{ignore} ) ? $opt->{ignore} : 0;
    my $limit = ( exists $opt->{limit} ) ? $opt->{limit} : 1024;

    my $w = scalar(@{$pixels_ref->[0]});
    my $h = scalar(@{$pixels_ref});

    my @figures = ();
    my $y0 = 0;
    foreach my $point_number (1..$limit) {
        my ( $x, $y ) = ( -1, -1 );

        for (my $iy=$y0; $iy<$h; $iy++) {
            my $i = 0;
            #say "iy = $iy";
            foreach my $val (@{$pixels_ref->[$iy]}) {
                if ( $val != $ignore ) {
                    #say "here!!!!!!!!!!!!!";
                    $x = $i;
                    $y = $iy;
                    last;
                }
                $i++;
            }

            last if 0 <= $y;
            $y0 = $iy + 1;
        }

        if ( $y < 0 ) {
            last;
        }

        my $x0 = $x;
        my $y0 = $y;
        my $trace_value = $pixels_ref->[$y][$x];

        # 探索開始点は図形の左下なので、
        # 上方向と繋がってなければ閉じていないと判断できる
        my $is_close = 0;
        if ( $x0 < ($w - 1) and $y0 < ($h - 1) ) {
            if ( $pixels_ref->[$y][$x + 1] == $trace_value
             and $pixels_ref->[$y + 1][$x] == $trace_value ) {
                $is_close = 1;
            }
        }

        if ( $is_close ) {
            # 最後の探索が開始点に到達できるように残しておく
        }
        else {
            # 閉じた図形ではないので、探索済みの処理をする
            $pixels_ref->[$y][$x] = $ignore;
        }

        my @points = ( [$x, $y] );
        my $search_comp = 0;
        while ( not $search_comp ) {
            my $number_of_points = scalar( @points );

            # 右方向に探索
            if ( ($x + 1) < $w and $pixels_ref->[$y][$x + 1] == $trace_value ) {
                while ( $pixels_ref->[$y][$x + 1] == $trace_value ) {
                    $x++;
                    $pixels_ref->[$y][$x] = $ignore;
                    last if $w <= ($x + 1);
                }

                push @points, [ $x, $y ];
            }

            # 上方向に探索
            if ( ($y + 1) < $h and $pixels_ref->[$y + 1][$x] == $trace_value ) {
                while ( $pixels_ref->[$y + 1][$x] == $trace_value ) {
                    $y++;
                    $pixels_ref->[$y][$x] = $ignore;
                    last if $h <= ($y + 1);
                }

                push @points, [ $x, $y ];
            }

            # 左方向に探索
            if ( 0 < ($x - 1) and $pixels_ref->[$y][$x - 1] == $trace_value ) {
                while ( $pixels_ref->[$y][$x - 1] == $trace_value ) {
                    $x--;
                    $pixels_ref->[$y][$x] = $ignore;
                    last if ($x - 1) < 0;
                }

                push @points, [ $x, $y ];
            }

            # 下方向に探索
            if ( 0 < ($y - 1) and $pixels_ref->[$y - 1][$x] == $trace_value ) {
                while ( $pixels_ref->[$y - 1][$x] == $trace_value ) {
                    $y--;
                    $pixels_ref->[$y][$x] = $ignore;
                    last if ($y - 1) < 0;
                }

                push @points, [ $x, $y ];
            }

            # 探索前と頂点を比較することで完了したか判定
            if ( $number_of_points == scalar(@points) ) {
                if ( $is_close ) {
                    my ( $p1, $p2 ) = @points[0,-1];
                    if ( $p1->[0] == $p2->[0] and $p1->[1] == $p2->[1] ) {
                        # 開始点と終点が同じなので、終点を取り除いて探索終了
                        pop @points;
                        $search_comp = 1;
                    }
                    else {
                        # 閉じていないことが判明したので、開始点から再探索
                        $is_close = 0;
                        @points = reverse @points;
                        $pixels_ref->[$y0][$x0] = $ignore;
                        ( $x, $y ) = ( $x0, $y0 );
                    }
                }
                else {
                    $search_comp = 1;
                }
            }
        }

        push @figures, +{
            points    => \@points,
            is_closed => $is_close,
            value     => $trace_value
        };
    }

    return \@figures;
}

1;
__END__

=encoding utf-8

=head1 NAME

Imager::LineTrace::Algorithm - Line trace algorithm

=head1 SYNOPSIS

    use Imager::LineTrace::Algorithm;

    my @pixels = Imager::LineTrace::Figure->new(
        [ 1, 1, 1 ],
        [ 1, 0, 1 ],
        [ 1, 1, 1 ]
    );

    my %args = ( ignore => 0 );
    my $figures_ref = Imager::LineTrace::Algorithm::search( \@pixels, \%args );

=head1 DESCRIPTION

Trace algorithm for Imager::LineTracer.

RETURN DATA

    # $figures_ref is ARRAY reference.
    my $figures_ref = Imager::LineTrace::Algorithm::search( \@pixels, \%args );

    # $figure_ref is HASH reference.
    my $figure_ref = $figures_ref->[0];

    # $figure_ref->{points} is ARRAY reference.
    foreach my $point (@{$figure_ref->{points}}) {
        printf( "x = %d, y = %d\n", $point->[0], $point->[1] );
    }

    # Traced pixel value.
    print $figure_ref->{value}, "\n";

    # Figure is closed.
    print $figure_ref->{is_closed}, "\n";

=head1 LICENSE

Copyright (C) neko.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

neko E<lt>techno.cat.miau@gmail.comE<gt>

=cut


lib/Imager/LineTrace.pm
package Imager::LineTrace;
use 5.008001;
use strict;
use warnings;

use base qw(Imager);

use Imager::LineTrace::Algorithm;
use Imager::LineTrace::Figure;

our $VERSION = "0.05";

sub line_trace {
    my $self = shift;
    my %args = @_;

    my $channels = [ 0, 1, 2 ];
    if ( exists $args{channels} ) {
        $channels = $args{channels};
    }

    my $number_of_channels = scalar( @{$channels} );
    my $ymax = $self->getheight() - 1;
    my @pixels = map {
        my $iy = $_;

        my $ary_ref = $self->getsamples( y => $iy, channels => $channels );
        my @wk = ();
        if ( @{$channels} == 3 ) {
            my @tmp = unpack( "C*", $ary_ref );
            while ( @tmp ) {
                my $val = shift @tmp;
                $val = ($val << 8) + shift @tmp;
                $val = ($val << 8) + shift @tmp;
                push @wk, $val;
            }
        }
        elsif ( @{$channels} == 2 ) {
            @wk = unpack( "S*", $ary_ref );
        }
        else {
            @wk = unpack( "C*", $ary_ref );
        }

        \@wk;
    } 0..$ymax;

    if ( not exists $args{ignore} ) {
        if ( @{$channels} == 3 ) {
            $args{ignore} = 0xFFFFFF;
        }
        elsif ( @{$channels} == 2 ) {
            $args{ignore} = 0xFFFF;
        }
        else {
            $args{ignore} = 0xFF;
        }
    }

    my $results = Imager::LineTrace::Algorithm::search( \@pixels, \%args );
    my @figures = map {
        Imager::LineTrace::Figure->new( $_ );
    } @{$results};

    return \@figures;
}

1;
__END__

=encoding utf-8

=head1 NAME

Imager::LineTrace - Line tracer

=head1 SYNOPSIS

    use Imager::LineTrace;

    my $img = Imager::LineTrace->new( file => $ARGV[0] ) or die Imager->errstr;
    my $figures_ref = $img->line_trace();

=head1 DESCRIPTION

    # Tracing clockwise from left top.

Expected Input and Result

    # Enter a figure made up of line vertical or horizontal.
    my $img = Imager::LineTrace->new( xsize => 16, ysize => 16 );
    $img->box( filled => 1, color => 'white' );
    $img->setpixel( x => 3, y => 2, color => '#000000' );
    $img->line( x1 => 6, y1 => 5,
                x2 => 9, y2 => 5, color => '#333333' );
    $img->polyline( points => [
            [ 2,  8 ],
            [ 5,  8 ],
            [ 5, 11 ]
        ], color => '#666666' );
    $img->box( xmin => 10, ymin => 10,
               xmax => 14, ymax => 14, color => '#999999' );

    my $figures_ref = $img->line_trace();

    # from Sample/bmp2figure.pl
    my $i = 0;
    foreach my $figure (@{$figures_ref}) {
        print "-------- [", $i++, "] --------", "\n";
        print "type        : ", $figure->{type}, "\n";
        print "trace_value : ", sprintf("0x%06X", $figure->{value}), "\n";
        print "is_close: ", $figure->{is_closed}, "\n";
        foreach my $p (@{$figure->{points}}) {
            printf( "(%2d,%2d)\n", $p->[0], $p->[1] );
        }
    }

    # -------- [0] --------
    # type        : Point
    # trace_value : 0x000000
    # is_closed   : 0
    # ( 3, 2)
    # -------- [1] --------
    # type        : Line
    # trace_value : 0x333333
    # is_closed   : 0
    # ( 6, 5)
    # ( 9, 5)
    # -------- [2] --------
    # type        : Polyline
    # trace_value : 0x666666
    # is_closed   : 0
    # ( 2, 8)
    # ( 5, 8)
    # ( 5,11)
    # -------- [3] --------
    # type        : Polygon
    # trace_value : 0x999999
    # is_closed   : 1
    # (10,10)
    # (14,10)
    # (14,14)
    # (10,14)

Basic Overview

    my $img = Imager::LineTrace->new( file => $path ) or die Imager->errstr;

    # Trace black line on white.
    my $figures_ref = $img->line_trace();

    # If you want to trace counter-clockwise from left bottom.
    $img->filp( dir => 'v' );
    my $figures_ref = $img->line_trace();

    # If you want to select color. ( 0:R, 1:G, 2:B, 3:Alpha )
    my $figures_ref = $img->line_trace( channels => [0] );

    # Or you want to trace with R,G and B.(this is defalt.)
    my $figures_ref = $img->line_trace( channels => [0,1,2] );

    # If you want to trace not black color.
    my $figures_ref = $img->line_trace( ignore => 0 );

    # If you want to trace many figure. (default "limit" is 1024)
    my $figures_ref = $img->line_trace( limit => 10000 );

=head1 LICENSE

Copyright (C) neko.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

neko E<lt>techno.cat.miau@gmail.comE<gt>

=cut

